parlai/core/torch_generator_agent.py

```diff
@@ -0,0 +1,1020 @@
+            scores = self.decode_forced(ys, encoder_states)
+        else:
+            bsz = xs.size(0)
+            scores = self.greedy_decode(
+                bsz, encoder_states,
+                maxlen or self.longest_label
+            )
+
+        return scores, encoder_states
+
+
+class TorchGeneratorAgent(TorchAgent):
+    """Abstract Generator agent. Only meant to be extended.
+
+    TorchGeneratorAgent aims to handle much of the bookkeeping and
+    infrastructure work for any generative models, like seq2seq or transformer.
+    It implements the train_step and eval_step. The only requirement is that
+    your model *must* implemented the interface TorchGeneratorModel interface.
+    """
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        agent = argparser.add_argument_group('Torch Generator Agent')
+        agent.add_argument('--beam-size', type=int, default=1,
+                           help='Beam size, if 1 then greedy search')
+        agent.add_argument('--beam-dot-log', type='bool', default=False, hidden=True,
+                           help='Dump beam trees as png dot images into /tmp folder')
+        agent.add_argument('--beam-min-n-best', type=int, default=3,
+                           help='Minimum number of nbest candidates to achieve '
+                                'during the beam search')
+        agent.add_argument('--beam-min-length', type=int, default=1,
+                           help='Minimum length of prediction to be generated by '
+                                'the beam search')
+        agent.add_argument('--beam-block-ngram', type=int, default=0, hidden=True,
+                           help='Block all repeating ngrams up to history length n-1')
+        agent.add_argument('--skip-generation', type='bool', default=False, hidden=True,
+                           help='Skip beam search. Useful for speeding up training, '
+                                'if perplexity is the validation metric.')
+
+        super(TorchGeneratorAgent, cls).add_cmdline_args(argparser)
+        return agent
+
+    def __init__(self, opt, shared=None):
+        init_model = None
+        if not shared:  # only do this on first setup
+            # first check load path in case we need to override paths
+            if opt.get('init_model') and os.path.isfile(opt['init_model']):
+                # check first for 'init_model' for loading model from file
+                init_model = opt['init_model']
+            if opt.get('model_file') and os.path.isfile(opt['model_file']):
+                # next check for 'model_file', this would override init_model
+                init_model = opt['model_file']
+
+            if init_model is not None:
+                # if we are loading a model, should load its dict too
+                if (os.path.isfile(init_model + '.dict') or
+                        opt['dict_file'] is None):
+                    opt['dict_file'] = init_model + '.dict'
```

Comment:
I kind of hate this code (I mean, I wrote it) but I'm not sure what else we should do with it / where else it should be exactly, maybe in a function in torch agent?